Book: Update QObjects chapter to v0.4 (#320)
KDAB/cxx-qt: book/src/qobject/cxxqtthread.md

# CxxQtThread

`CxxQtThread<T>` is used for easy [threading](../concepts/threading.md) with QObjects.
The QObjects generated by CXX-Qt are neither [`Sync`](https://doc.rust-lang.org/std/marker/trait.Sync.html) nor [`Send`](https://doc.rust-lang.org/std/marker/trait.Send.html).
Therefore they may not be passed between threads.

> `CxxQtThread<T>` is a handle to the Qt thread that the `qobject::T` lives on.

Reviewer: "Handle" doesn't seem like quite the right word here; that makes it seem like it's similar to a [std::thread::JoinHandle](https://doc.rust-lang.org/std/thread/struct.JoinHandle.html). Maybe "helper for running Rust code in the Qt event loop"? That's a bit verbose though :/

Author: Hm, true, it's definitely not a JoinHandle.
Maybe just calling it "reference" to the Qt thread would work?

Reviewer: "Reference" already has different meanings in Rust and C++, so I think that would be worse than "handle".

Reviewer 2: Maybe just "`CxxQtThread<T>` is an object that allows you to queue closures safely onto the Qt thread that `qobject::T` lives on."

Author: Done

This handle is `Send` and can therefore be moved into other threads.
It allows you to queue events from a different thread to occur on the thread of the `qobject::T` by using the Qt Event Loop.

To access the `CxxQtThread<T>` use the `qt_thread(&self)` method on a [`qobject::T`](./cpp_object.md).

```rust,ignore,noplayground
{{#include ../../../examples/qml_features/rust/src/threading.rs:book_qt_thread}}
```

The `CxxQtThread<T>` can then be moved into any Rust thread.

> The `queue(impl FnOnce(ctx: Pin<&mut TQt>)) -> bool` is used to queue a Rust function pointer onto the Qt event loop. The first argument in the function pointer is a pinned pointer to the C++ side of the QObject.

Reviewer: What is the meaning of the returned bool?

Author: Updated, it actually returns a Result now.
Basically, whether it was possible to queue the lambda or not.
Most of the time a failure means the QObject doesn't exist anymore.

# CxxQtThread

`CxxQtThread<T>` is used for easy [threading](../concepts/threading.md) with QObjects.

> The QObjects generated by CXX-Qt are neither [`Sync`](https://doc.rust-lang.org/std/marker/trait.Sync.html) nor [`Send`](https://doc.rust-lang.org/std/marker/trait.Send.html).

Reviewer: ```suggestion
The QObjects generated by CXX-Qt are neither [`Send`](https://doc.rust-lang.org/std/marker/trait.Send.html) nor [`Sync`](https://doc.rust-lang.org/std/marker/trait.Sync.html).
```
match the structure of the following sentence

Reviewer 2: done

> Therefore they may not be passed between threads.

Reviewer: ```suggestion
Therefore they may not be passed between threads nor accessed from multiple threads.
```

`CxxQtThread<T>` is an object that allows you to queue closures safely onto the Qt thread that the `qobject::T` lives on.
This object is `Send` and can therefore be moved into other threads.
It allows you to queue events from a different thread to occur on the thread of the `qobject::T` by using the Qt Event Loop.

To access the `CxxQtThread<T>` use the `qt_thread(&self)` method on a [`qobject::T`](./generated-qobject.md).

```rust,ignore,noplayground
{{#include ../../../examples/qml_features/rust/src/threading.rs:book_qt_thread}}
```

The `CxxQtThread<T>` can then be moved into any Rust thread.

> The `queue` function can then be used to queue a lambda onto the Qt event loop:

Reviewer: ```suggestion
The `queue` function can then be used to queue a [closure](https://doc.rust-lang.org/book/ch13-01-closures.html) onto the Qt event loop:
```

Reviewer: I haven't seen the term "lambda" used much in the context of Rust.

Reviewer 2: done

``` rust,ignore,noplayground
fn queue(&self, f: F) -> Result<(), cxx::Exception>
  where F: impl FnOnce(ctx: Pin<&mut TQt>) + Send + 'static
```

> The first argument of the lambda is a pinned mutable reference to the `qobject::T`.

Reviewer: ```suggestion
The first argument of the closure is a pinned mutable reference to the `qobject::T`.
```

Reviewer 2: done
